olm/operator-registry.md

```diff
@@ -0,0 +1,291 @@
+
+The format then looks something like the operator-registry database that exists today, but with only minimal information about every single container. The db schema will have a layer of "latest" for each operator name for all the metadata that drives the OpenShift console (descriptions, icon references, etc.). Then all operator image pointers will aggregate just the metadata that drives install and upgrade (version, channels, dependencies, owned crds). To add additional updates to the database, the database just needs to be inserted/batch updated with changes rather than rebuilt from scratch.
+
+---------------------
+
+_**Serve Operator Database**_
+
+As an Operator developer, I want my Operator to be consumed and made available to clusters.
+
+- So that the metadata attached to my operator can be used to determine what is available to install on a cluster.
+
+Acceptance Criteria:
+I can create a registry database that hosts a history of my operator and other developer's operators. This database can be used by openshift clusters.
+
+---------------------
+
+_**Add Indexes to cluster**_
+
+As an OpenShift administrator, I want to have a way of adding an operator index to my cluster
+
+- So that I can make operators available to install onto my cluster
+
+Acceptance Criteria:
+When I add an operator index to my cluster, I am able to get information about the operators defined in that index.
+
+Internal detail note:
+Create an image type CatalogSource that, when added to a cluster with a reference to an index image, pulls the index manifest down and generates the operator metadata database for that operator.
+
+---------------------
+
+_**Drive Operator Installs and UX**_
+
+As an OpenShift administrator, I want a way to install operators made available to me from my Operator index.
+- So that I can view what operators are available to install
+- So that I can view metadata about that operator (description, author, etc) on the OpenShift console
+- So that I can select a given operator name, version, and channel to subscribe to
+
+Acceptance Criteria:
+
+When I install an operator index onto my cluster, I can use the data that is available on the cluster to drive operator installation and lifecycle.
+
+---------------------
+
+_**Polling Updates**_
+
+As an OpenShift administrator, I want a way to configure a poll timer to ensure my index of operators is up to date and references all of the latest operator bundles.
+
+- So that I can get updates over time
+- So that I can know the frequency that those updates will appear
+
+Acceptance Criteria:
+When I set the update frequency on an operator index, OLM will poll and ensure that if there is a new version of the index that those changes are pulled and propagated into the cluster.
+
+Internal details:
+Polling process should be reviewed by quay team to ensure we scale on quay (although these indexes should work on ANY container image registry)
```

Comment:
What will this polling look like? Are you proposing that catalog-operator in OLM would be reaching out to an image registry directly to see if new versions of the index image are available?

This feels like scheduled imports on ImageStreams. @bparees thoughts here?

Reply:
I've had a few conversations with @shawn-hurley about the challenges involved in pulling images directly.  Most of those concerns would also apply to anything that's checking an external registry for images as well.  Off the top of my head that is:

1) proxy support
2) credential management (this is the big one... ideally you want to benefit from creds the admin already setup for the cluster, not force people to point you to yet another set of creds that this particular component needs)
3) CA trust management (for registries that are signed with self-signed certs or CAs that aren't in system trust)
(if you're actually pulling images yourself, there are additional concerns primarily around storage management).


So those are all things to think about here.

But as to your specific question, yes it sounds like imagestream scheduled import...  if OLM pre-req'd openshift, it would be pretty reasonable to simply setup imagestreams with scheduled import turned on, and then use a form of imagechange trigger to have OLM take action when the imagestream picks up a new SHA.

But of course if OLM is expected to work on k8s, that's not viable, and it sounds like this polling implementation will have to rebuild the capabilities that exist in imagestream import.